### PR TITLE
feat: add model fallback support for resilient AI in DDIL environments

### PR DESCRIPTION
**Pull request by crspeller**
#### Summary

Add a service-level model fallback mechanism so AI automatically switches between LLM providers when the primary is unavailable — using cloud LLMs when connected and falling back to on-prem/local models in DDIL (Denied, Disrupted, Intermittent, Limited) environments.

**How it works:**

Each AI service can now optionally specify a **Fallback Service** — another configured service to try when the primary fails (network error, rate limit, model unavailable). Fallback chains are supported (e.g., OpenAI → Anthropic → Local Ollama), with automatic cycle detection.

This leverages Bifrost's built-in fallback mechanism, which handles sequential failover automatically on every request. When the primary provider recovers, it's automatically used again — no manual intervention needed.

**Key behavior:**
- Automatic (no user action required)
- Graceful degradation (degraded capability instead of total loss)
- Per-request failover (not sticky — primary is always tried first)
- Minimal disruption to workflow

**Backend changes:**
- Added `FallbackServiceID` field to `ServiceConfig` for chaining services
- Added `ResolveFallbackChain()` with cycle detection (visited-set algorithm)
- Created `multiProviderAccount` to register multiple providers with Bifrost
- Updated Bifrost request builders to attach fallback chain to every request
- Updated bot initialization to resolve and pass fallback chains
- Updated optimistic change detection to include fallback service configs

**Frontend changes:**
- Added "Fallback Service" dropdown to the service configuration panel
- Self-selection excluded from dropdown to prevent trivial cycles
- Help text explains the feature and chain support

**Tests added (28 new test cases):**
- `ResolveFallbackChain`: no fallback, simple chain, multi-hop chain, cycle detection, self-cycle, missing/invalid services
- `multiProviderAccount`: single/multiple providers, unknown provider, duplicate handling, Azure/Bedrock configs
- Fallback request building: Chat API and Responses API with/without fallbacks
- `NewFromServiceConfig`: no fallbacks, single/multiple fallbacks, bot model override isolation
- `serviceConfigToFallbackEntry`: provider mapping, URL normalization

#### Ticket Link

N/A — Feature request for resilient AI in DDIL environments

#### Release Note
```release-note
Added model fallback support for AI services. Each service can now specify a fallback service that will be used automatically when the primary provider is unavailable. Fallback chains are supported (e.g., cloud → on-prem → local). This enables graceful degradation instead of total AI loss in DDIL (disconnected) environments.
```

<div><a href="https://cursor.com/agents/bc-8c77cc0d-96e6-45dd-98db-9a6400af1bbe"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-8c77cc0d-96e6-45dd-98db-9a6400af1bbe"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

## Summary by CodeRabbit

* **New Features**
  * Ordered multi-provider fallback chains for automatic failover, with Responses→chat downgrades for chat-only fallbacks and support for chained fallback services.

* **Bug Fixes**
  * Credential redaction now consistently scrubs multiple provider keys from logs and errors.

* **Tests**
  * Expanded unit and end-to-end tests covering fallback wiring, multi-hop failover, and configuration edge cases.

* **Localization**
  * Added English strings for the fallback service UI.

### CODE REVIEW

File: bifrost/config.go [hunk 1]
```diff
@@ -124,9 +127,67 @@ func NewFromServiceConfig(serviceConfig llm.ServiceConfig, botConfig llm.BotConf
 		cfg.DefaultModel = botConfig.Model
 	}
 
+	// Build fallback entries from the resolved fallback service chain.
+	for _, fbSvc := range fallbackServices {
+		fbEntry, fbErr := serviceConfigToFallbackEntry(fbSvc)
+		if fbErr != nil {
+			// Skip unsupported fallback services rather than failing the whole bot.
+			continue
```
Comment (nickmisasi) [inline bifrost/config.go]: I think silently skipping is a bad idea if you consider the audience for this feature. We don't want to give someone the illusion of a fallback being in place for it to not work when its needed

File: bifrost/bifrost.go [hunk 2]
```diff
@@ -99,11 +105,62 @@ type Config struct {
...
+	// local Ollama server). Only consulted when the fallback is registered as a
+	// custom provider.
+	IsKeyLess bool
+	// ChatOnly marks an OpenAI-base fallback whose endpoint does not support the
+	// Responses API (e.g. a local Ollama/vLLM server). Such a fallback is
+	// registered as a custom provider with chat-only AllowedRequests so Bifrost
+	// downgrades a Responses-API request to chat completions for it.
+	ChatOnly                bool
+	StreamingTimeoutSeconds int
 }
```
Comment (nickmisasi) [inline bifrost/bifrost.go:135]: `FallbackEntry` duplicates a lot of the Config. Can this be simplified? This might be one of the first times I've seen a use case for a LinkedList IRL (`NextFallback *Config`) 

At the very least we should have a shared type for the provider config fields, or use struct embedding.

File: bifrost/bifrost_test.go [hunk 3]
```diff
@@ -1223,6 +1224,927 @@ func TestConvertToBifrostResponsesRequestStructuredOutput(t *testing.T) {
...
+		"a same-base fallback must not collide on the primary's provider slot")
+	assert.NotEqual(t, llmInstance.fallbacks[0].Provider, llmInstance.fallbacks[1].Provider,
+		"each fallback must occupy a distinct slot")
+	assert.Equal(t, "llama3", llmInstance.fallbacks[1].Model)
+}
+
+// TestNewFromServiceConfig_SkipsUnmappableFallbackInChain pins the resilience
+// contract that a fallback service which cannot be mapped to a Bifrost provider
+// is skipped rather than failing the whole bot — the whole point of a DDIL
+// failover feature is to degrade, not to break. A regression that returned the
```
Comment (nickmisasi) [inline bifrost/bifrost_test.go]: I think again related to that silent skip, I sort of see the logic but I think it's backwards. You would want the service to degrade immediately and obviously on setup rather than when its mission-critical to work

File: bifrost/bifrost.go [hunk 4]
```diff
@@ -99,11 +105,62 @@ type Config struct {
...
 
 // providerAccount implements the Bifrost Account interface for a single provider.
 type providerAccount struct {
-	provider                schemas.ModelProvider
+	// provider is the base Bifrost provider type (e.g. schemas.OpenAI), used to
+	// build provider-specific key and network configuration.
+	provider schemas.ModelProvider
```
Comment (nickmisasi) [inline bifrost/bifrost.go]: LLM went a little overboard on comments in this file

**Review by nickmisasi — CHANGES_REQUESTED** [review]
Just a couple flags

**Review by nickmisasi — APPROVED** [review]
